### PR TITLE
Implement Authorization Code Flow from Spotify API

### PR DESCRIPTION
**Pull request by ncbecker**
More infos on Authorization with Spotify API:
https://developer.spotify.com/documentation/general/guides/authorization-guide/#authorization-code-flow

## Related Issue
#34

### CODE REVIEW

**Review by phlgr — APPROVED** [review]
Great Implementation

**Comment by ncbecker** [comment]
The deployment throws errors, dotenv variables are not read, although they are added as Config Vars for the PR and main in Heroku. It's working in the dev environment. Anyone an idea?

File: server.js [hunk 1]
```diff
@@ -1,20 +1,143 @@
...
+    } = await getAppAccessToken();
+    response.cookie("apptoken", newAppToken, {
+      maxAge: (expiresIn - 60) * 1000,
+    });
+    const episodeData = await getEpisodeInfo(newAppToken, id);
+    response.send(episodeData);
+    return;
+  }
+  const episodeData = await getEpisodeInfo(accessToken, id);
   response.send(episodeData);
```
Comment (AlexisRoe) [inline server.js]: double of code, replace it with one function , which will do that, or think about using it once with a changing const

```suggestion
    const episodeData = await getEpisodeInfo(newAppToken, id);
    response.send(episodeData);
    return;
  }
  const episodeData = await getEpisodeInfo(accessToken, id);
  response.send(episodeData);
```

something like that 

`````js
const episodeData = await getEpisodeInfo(newAppToken || accessToken, id);
  response.send(episodeData);

`````

File: server.js [hunk 2]
```diff
@@ -1,20 +1,143 @@
...
+  const refreshToken = request.cookies.refresh;
+  if (!accessToken && !refreshToken) {
+    response.redirect("/oauth/spotify/authorize");
+    return;
+  }
+  if (!accessToken) {
+    const {
+      access_token: newAccessToken,
+      expires_in: expiresIn,
+      refresh_token: newRefreshToken,
```
Comment (AlexisRoe) [inline server.js]: try to give the same Variables back you use here or rewrite the function to directly use the variables
```suggestion
      access_token,
      expires_in,
      refresh_token
```

File: server.js [hunk 3]
```diff
@@ -1,20 +1,143 @@
...
+    } = await refreshUserAccessToken(refreshToken);
+
+    response.cookie("access", newAccessToken, {
+      maxAge: (expiresIn - 60) * 1000,
+    });
```
Comment (AlexisRoe) [inline server.js]: did you use, 3 different cookies? in /api/episode/:id you use/ set a cookie named "apptoken", here you set a cookie named access... in line 94/95 is the third one

````js
const accessToken = request.cookies.access;
  const refreshToken = request.cookies.refresh;
````

File: server.js [hunk 4]
```diff
@@ -1,20 +1,143 @@
...
+    }
+    const userProfileData = await getCurrentUserProfile(newAccessToken);
+    response.send(userProfileData);
+    return;
+  }
+  try {
+    const userProfileData = await getCurrentUserProfile(accessToken);
+    response.send(userProfileData);
+  } catch (error) {
+    console.error(error);
```
Comment (AlexisRoe) [inline server.js]: you do the same twice, once with try/catch the other you don't use

```suggestion
try {
    const userProfileData = await getCurrentUserProfile(newAccessToken);
    response.send(userProfileData);
    return;
    } catch (error) {
    console.error(error);
  }
  try {
    const userProfileData = await getCurrentUserProfile(accessToken);
    response.send(userProfileData);
  } catch (error) {
    console.error(error);
```

**Review by AlexisRoe — COMMENTED** [review]
good job of implementing the oauth

File: lib/SpotifyMethods.js [hunk 5]
```diff
@@ -1,5 +1,36 @@
 const fetch = require("node-fetch");
 
+async function getCurrentUserProfile(token) {
+  console.log(token);
```
Comment (marwinburesch) [inline lib/SpotifyMethods.js]: 🧐

File: server.js (hunk 2 above)
Comment (marwinburesch) [reply server.js]: I actually like the renaming :)

File: server.js [hunk 6]
```diff
@@ -1,20 +1,143 @@
...
+  }
```
Comment (marwinburesch) [inline server.js]: I think you should wrap the whole block in `try... catch`. What if the Spotify API is down or something goes wrong during the auth process?

**Review by marwinburesch — CHANGES_REQUESTED** [review]
I don't know much about the Spotify OAuth Flow, but this looks nice! 🐜 
You got so many Wrappers, it's basically a gang 🦛 

RC cos of the try catch
Cheers, 🍻

**Comment by ncbecker** [comment]
Thanks @phlgr @AlexisRoe @marwinburesch for your reviews and comments 🙏! Most points are addressed in PR https://github.com/ncbecker/podium/pull/55.
I know I have too many wrappers 👀 this definitely needs refactoring (for now I am happy it works)

**Pull request merged**